certbot/interfaces.py

```diff
@@ -604,3 +604,204 @@ def renew_deploy(self, lineage, *args, **kwargs):
         :type lineage: storage.RenewableCert
 
         """
+
+
+@six.add_metaclass(abc.ABCMeta)
+class ParserNode(object):
+    """
+    ParserNode is the basic building block of the tree of such nodes,
+    representing the structure of the configuration. It is largely meant to keep
+    the structure information intact and idiomatically accessible.
+
+    The root node as well as the child nodes of it should be instances of ParserNode.
+    Nodes keep track of their differences to on-disk representation of configuration
+    by marking modified ParserNodes as dirty to enable partial write-to-disk for
+    different files in the configuration structure.
+
+    While for the most parts the usage and the child types are obvious, "include"-
+    and similar directives are an exception to this rule. This is because of the
+    nature of include directives - which unroll the contents of another file or
+    configuration block to their place. While we could unroll the included nodes
+    to the parent tree, it remains important to keep the context of include nodes
+    separate in order to write back the original configuration as it was.
+    """
+
```

Comment:
Will a `ParserNode` (or any of its subclasses) have any public attributes? If so, I think we should define them using an approach like: https://stackoverflow.com/questions/2736255/abstract-attributes-in-python

Reply:
Good point. I'll add those. Examples of public attributes would be `ancestor` and `children`.